### ISSUE TITLE
Bug: MissedProviderError in `.backward()` with multiple invokers

### ISSUE THREAD

**Issue by lmjantsch** [ndif-team/nnsight#664]
# Bug: MissedProviderError in `.backward()` with multiple invokers

## Reproducing example

```python
with model.trace() as tracer:
    with tracer.invoke(inp_str):          # invoker 1
        out = model.lm_head.output.save()

    with tracer.invoke(inp_str):          # invoker 2
        attn_outs = []
        for layer in model.transformer.h:
            x = layer.attn.c_proj.output
            attn_outs.append(x)

        logits = model.lm_head.output
        loss = logits.sum()

        with loss.backward():
            for x in reversed(attn_outs):
                print(x.grad)            # MissedProviderError here
```

Works fine when only invoker 2 is present (the `.backward()` block is removed or invoker 1 is dropped).

## What goes wrong

With a **single invoker**, `Batcher.narrow()` returns the original module output tensor unchanged (`needs_batching = False`). That original tensor is part of the computation graph that produced `loss`, so PyTorch runs through it during `loss.backward()` and the hook fires.

With **two or more invokers** sharing an input, `needs_batching = True` and `Batcher.narrow()` returns `acts.narrow(0, batch_start, batch_size)` — a **storage-sharing view** of the original tensor. This view is what gets stored in `attn_outs`.

The backward hook is registered on this view. However, the view is **never used as an input to any operation that produced `loss`**. The computation graph for `loss` flows through the original `c_proj_output` tensor (the full-batch output produced during the forward pass), not through the narrowed view that was created purely to deliver a batch slice to the worker thread.

Because the view has no path in `loss`'s backward graph, PyTorch never runs a gradient node for it. The hook never fires.

## Proposed fix

Before registering the backward hook, check whether the tensor is a **storage-sharing view** (i.e., `tensor._base is not None`). If it is, derive the view's position in the parent tensor from standard tensor metadata — storage offset, stride, and size along the batch dimension — and register the hook on the **parent tensor** instead. When the hook fires, narrow the incoming gradient to the correct slice before delivering it to the `BackwardsMediator`. If the user modified the gradient (a swap event), splice the modified slice back into the full parent gradient before returning it to PyTorch.

The provider/requester key (`id(tensor)`) stays tied to the **original view object**, so the `BackwardsMediator`'s existing wait logic requires no changes.

## Open question / potential issue

The fix assumes that whenever the tensor delivered to the worker is a view, the hook should be on the **parent** (base tensor). This is correct for the `Batcher.narrow()` case, because the view was created only for slice delivery and was never fed into a computation leading to `loss`.

However, consider a case where the user themselves constructs `loss` from a narrowed view:

```python
with tracer.invoke(inp_str):
    m_out_last_token = model.transformer.h[-1].output.narrow(1, -1, 1) # select last token hidden_state
    loss1 = loss_fn1(m_out_last_token)

    logits = model.lm_head.output
    loss2 = loss_fn2(logits)

    loss = loss1 + loss2

    with loss.backward():
        print(logits.grad) # works fine
        print(m_out_last_token.grad) # would get the gradient from model.transformer.h[-1].output which is the combined partial gradient through both losses.
```
